Improve innerJoin performance

DanielMao1 opened the pull request:
Thanks for contributing.

- [x] Tick to sign-off your agreement to the [Developer Certificate of Origin (DCO) 1.1](https://developercertificate.org)

## Description

Fix #902 
1. `tech.tablesaw.joining.DataFrameJoiner#joinInternal` method
I add the logic to determine which table will be used to do row iteration when table1 and table2 have different sizes. If table1 has a larger size than table B, table2 will be used to do row iteration. It significantly improves performance when table1 is much larger than table2.
2.`tech.tablesaw.joining.DataFrameJoiner#createMultiColSelection` method
This method is to choose the columns of a table used as the join key.  Originally, only the left table is chosen, **while I add the method parameters `List<Integer> joinColumnIndexes`, so both tables can be chosen to generate the col key index to do join.** 
## Testing
Did you add a unit test?
I added a test case called `tech.tablesaw.joining.DataFrameJoinerPerformanceTest#innerJoinCustomersFirstTimeMeasure`, and you can use this case to compare the performance between big table join the small table and small table join big table.

DanielMao1 commented:
I sign up the agreement now @benmccann ,too. Thanks.

lwhite1 commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -355,10 +385,21 @@ private Selection selectionForColumn(Column<?> valueColumn, int rowIndex, Index
+   *
+   * @param table1 the table which used to generate Selection.
+   * @param ri row number of row in table 1.
+   * @param selectionSize max size in table 1.
```
It seems to be missing the description of argument 
   
    List<Index> indexes

I don't understand what the description of joinColumnIndexes is telling me. It appears that it should read
   the column indexes (plural) of table1
but what column indexes, and what are they used for?

DanielMao1 replied:
Thanks for your review:) 

> but what column indexes, and what are they used for?


It is the index in tables' column list, representing "join key" for joining. and will be used for cross join and catcat table.

> I don't understand what the description of joinColumnIndexes is telling me. It appears that it should read
the column indexes (plural) of table1


Before I change, it would read the column index of table1; 
After I change, It can be the column index of table2 for better performance. **I will change the variable name and JAVA doc tomorrow for more clear understanding.**

lwhite1 commented:
Hi @DanielMao1. Thanks for the PR. Did you demonstrate
a) this works on multi-table joins? This may be proven by existing tests, but it would be nice to add a test that we know exercises the new code.
b) the resulting table's column order is the same as it would be without this change?  It seems the fix only changes the row iteration order, but i just want to confirm that the resulting table is identical - at least in terms of column order, to what you would get prior to the change (i.e., table 1's columns appear in order, followed by table two's columns)?

DanielMao1 commented:
> Hi @DanielMao1. Thanks for the PR. Did you demonstrate
> a) this works on multi-table joins? This may be proven by existing tests, but it would be nice to add a test that we know exercises the new code.
> b) the resulting table's column order is the same as it would be without this change? It seems the fix only changes the row iteration order, but i just want to confirm that the resulting table is identical - at least in terms of column order, to what you would get prior to the change (i.e., table 1's columns appear in order, followed by table two's columns)?

Thanks for your response! 
a) yeah. Thanks for your advice. I will add other tests later for multiple joins.
b) It won't change the order of columns. the join order is determined in `crossJoin` method. I did not change such method.
The result table is identical, I will add some tests to prove it later.

benmccann commented on core/src/main/java/tech/tablesaw/joining/DataFrameJoiner.java:
```diff
@@ -198,31 +199,60 @@ private Table joinInternal(
-
-      Selection table1Rows = createMultiColSelection(table1, ri, table1Indexes, table1.rowCount());
-      Selection table2Rows = createMultiColSelection(table1, ri, table2Indexes, table2.rowCount());
+    // 2021/04/24 @Daniel Modified: If size of table1 is larger than table2,
```
I would remove "2021/04/24 @Daniel Modified:" from the comment. That's what Git history is for

DanielMao1 replied:
Thanks for your review: ) I deleted the comment and push another commit. Please check for the latest version.

DanielMao1 commented:
@lwhite1 @benmccann  
Hello: )
For the latest commit, I: 
###1.added two tests for the larger left table join smaller right table.
[test1](https://github.com/jtablesaw/tablesaw/blob/3d36d4fdd09e964f8b8a97d5d44cc7250ba547f4/core/src/test/java/tech/tablesaw/joining/DataFrameJoinerTest.java#L826) is for single join key and [test2](https://github.com/jtablesaw/tablesaw/blob/3d36d4fdd09e964f8b8a97d5d44cc7250ba547f4/core/src/test/java/tech/tablesaw/joining/DataFrameJoinerTest.java#L892) is for multiple join keys. It tests the column order and final result row number. 
**It shows that the code I modified does not influence the order of columns or multiple join results.**
### 2. fix the comments and Javadoc
If there are other inappropriate codes in the commits, I am looking forward to your suggestions.
Thanks.

lwhite1 commented on core/src/test/java/tech/tablesaw/joining/DataFrameJoinerPerformanceTest.java:
```diff
@@ -62,6 +62,26 @@ public void innerJoinCustomersFirst() {
   }
 
+  @Test
+  public void innerJoinCustomersFirstTimeMeasure() {
```
I'm not really sure what to do with this test, but I think you should disable it. We don't want to write (more) text strings to System.out. I spent a lot of time removing many that I put in myself. I would suggest using assertions rather than print statements, but I'm concerned that they will fail intermittently given the non-deterministic nature of performance tests. 

As long as you demonstrate correctness, and we "know" it improves performance, we don't need to run this every time. Also, any test that takes a long time to run is likely to break the CI runs.

DanielMao1 replied:
Yeah. I agree. I do not want to use assertion to do non-deterministic tests either. It is unstable.
I will delete this test in my PR. Maybe In the future we can find better ways to do performance improvement and test.
Thank you for your careful review : )

lwhite1 commented:
Thanks for the changes, @DanielMao1. I think if you disable the performance test, we're good to go.

DanielMao1 commented:
> Thanks for the changes, @DanielMao1. I think if you disable the performance test, we're good to go.

@lwhite1 I disabled my tests in the performance test.  I will only keep them in my local repository for further performance improvement in the future.  Thanks for your review.
